[HB-6261] Update the README.md of the Reference adapter with adapter creation guidelines (#50)
ChartBoost/chartboost-mediation-ios-adapter-reference: README.md

pod 'ChartboostMediationAdapterReference'
```

## Chartboost Mediation Custom Adapter Implementation Guide

> [!IMPORTANT]

> > Chartboost Mediation will not be providing official support for any custom adapters. For official adapters that we develop and support, [visit this site](https://adapters.chartboost.com).

Reviewer: ```suggestion
> Chartboost Mediation does not provide official support for custom adapters. For a list of official adapters, [visit this site](https://adapters.chartboost.com).
```

> 1. Create a new class and conform to Helium’s `PartnerAdapter` protocol.

Reviewer: ```suggestion
1. Create a new class and conform to Chartboost Mediation's `PartnerAdapter` protocol.
```

> 2. Implement var `partnerSDKVersion: String { get }` with the version of the partner SDK. Most adapters derive this from the partner SDKs API so that it always reports the correct number, even when a newer version is imported.

Reviewer: ```suggestion
2. Implement `var partnerSDKVersion: String { get }` with the version of the partner SDK. Most adapters derive this from the partner SDKs API so that it always reports the correct number, even when a newer version is imported.
```

> 3. Implement `var adapterVersion: String { get }` with the version of the mediation adapter. The adapter version format is `<Chartboost Mediation major version>.<Partner major version>.<Partner minor version>.<Partner patch version>.<Partner build version>.<Adapter build version>`.  

Reviewer: ```suggestion
3. Implement `var adapterVersion: String { get }` with the version of the Mediation adapter. The adapter version format is `<Chartboost Mediation major version>.<Partner major version>.<Partner minor version>.<Partner patch version>.<Partner build version>.<Adapter build version>`.  
```

`<Partner build version>` is optional, and omitted by most partners."

> For example, if this adapter is compatible with Helium SDK 4.x and partner SDK 1.2.3.x, and this is its initial release, then adapterVersion is 4.1.2.3.x.0.

Reviewer: ```suggestion
For example, if this adapter is compatible with Chartboost Mediation SDK 4.x and partner SDK 1.2.3.x, and this is its initial release, then adapterVersion is 4.1.2.3.x.0.
```

> 4. Implement var `partnerIdentifier: String { get }` with the internal identifier that the Helium SDK can use to refer to the current partner. Must match the value used in Chartboost Mediation's backend.

Reviewer: ```suggestion
4. Implement `var partnerIdentifier: String { get }` with the internal identifier that the Chartboost Mediation SDK can use to refer to the current partner. Must match the value used on the Chartboost Mediation dashboard.
```

I think the custom adapter identifier will be input by the publisher on the dashboard, but we should confirm this.

> 5. Implement `var partnerDisplayName: String { get }` the partner name as it should appear in text.

Reviewer: ```suggestion
5. Implement `var partnerDisplayName: String { get }`, the human readable partner name.
```

6. Implement `func setGDPR(applies: Bool?, status: GDPRConsentStatus)`,  
`func setCCPA(hasGivenConsent: Bool, privacyString: String)`,  
and `func setCOPPA(isChildDirected: Bool)`,  
which receive privacy settings from Chartboost Mediation SDK and apply them to the partner SDK.  
They are always called at startup when Chartboost Mediation initalizes adapters. They will also be called any time the publisher updates privacy settings on the Mediation SDK.  

> These privacy methods aren't called until *after* `setUp()`, so if your SDK requires privacy settings to be passed in at init then you will need to use default values on the very first launch and store the settings received by these methods for use on subsequent startups.

Reviewer: ```suggestion
These privacy methods aren't called until *after* `setUp()`, so if your SDK requires privacy settings to be passed in at init, then you will need to use default values on the very first launch, and store the settings received by these methods for use on subsequent startups.
```

Author: Neither of those commas feel right to me. I don't remember enough formal grammar to explain why, but fwiw Grammarly says the original is ok and that the added comma after "launch" is wrong (but it seems fine with or without the one following "init" so... can both ways be right?)

I suspect that something about the way I wrote this line feels off to you and the commas were an attempt to make it clearer. If you can put your finger on what's awkward about the current phrasing I'll try a different way of saying it.

Reviewer: Yeah, I'm guessing it's that both ways are technically grammatically correct (minus the second comma in my version). IMO the original version just feels too long to read all at once without being broken up by at least one comma.

Author: I broke it up into three sentences in [this new commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-reference/pull/50/commits/57477d26ab012578685085a6523730dbc9ccb1f2).

> 7. Implement `func setUp( with configuration: PartnerConfiguration, completion: @escaping (Error?) -> Void )` to initialize the partner SDK and perform any necessary setup in order to request and serve ads. If the operation succeeds, call `completion(nil)`. Otherwise, call `completion(Error)`.

Reviewer: ```suggestion
7. Implement `func setUp(with configuration: PartnerConfiguration, completion: @escaping (Error?) -> Void )` to initialize the partner SDK and perform any necessary setup in order to request and serve ads. If the operation succeeds, call `completion(nil)`. Otherwise, call `completion(Error)`.
```

If a call to this method times out, the Mediation SDK will consider your adapter un-initialized even if you later report a successful init.

> 8. Implement `func fetchBidderInformation( request: PreBidRequest, completion: @escaping ([String: String]?) -> Void )` to compute and call `completion([String: String])` of biddable token Strings. If network bidding is not supported by the current partner, return an empty Dictionary.

Reviewer: ```suggestion
8. Implement `func fetchBidderInformation( request: PreBidRequest, completion: @escaping ([String: String]?) -> Void )` to compute and call `completion([String: String])` of biddable token Strings. If network bidding is not supported by the current partner, call `completion(nil)`
```

9. Implement a class that wraps instances of your ads. It must conform to the `PartnerAd` protocol:  
    a. `var adapter: PartnerAdapter { get }`, which stores the adapter that created the ad.  
    b. `var request: PartnerAdLoadRequest { get }`, the associated ad load request.  
    c. `var delegate: PartnerAdDelegate? { get }`, the lifecycle events delegate.  

>     d. `var inlineView: UIView? { get }` View for displaying banner ads. Not used for other ad types.  

Reviewer: ```suggestion
    d. `var inlineView: UIView? { get }`: `UIView` that contains a banner ad, or `nil` for non-banner ads.
```

It also needs to store the `PartnerAdDelegate` that Chartboost Mediation SDK will pass to you, and call the following delegate methods to report your ad lifecycle events:  
    a. `func didTrackImpression(_ ad: PartnerAd, details: PartnerEventDetails)` when the partner SDK registers an impression for the currently showing ad.  
    b. `func didClick(_ ad: PartnerAd, details: PartnerEventDetails)` when the partner ad has been clicked as the result of a user action.  
    c. `func didReward(_ ad: PartnerAd, details: PartnerEventDetails)` when a reward has been given for watching a video ad.  
    d. `func didDismiss(_ ad: PartnerAd, details: PartnerEventDetails, error: Error?)` when the partner ad has been dismissed as the result of a user action.
    e. `func didExpire(_ ad: PartnerAd, details: PartnerEventDetails)` when the partner ad has expired as determined by the partner SDK.
10. Implement `func makeAd(request: PartnerAdLoadRequest, delegate: PartnerAdDelegate) throws -> PartnerAd`, which returns your `PartnerAd`-conforming objects.

> 11. In your `PartnerAd` ap the ad lifecycle events reported by your SDK to the appropriate `PartnerAdDelegate` methods on the 

Reviewer: ```suggestion
11. In your `PartnerAd` map the ad lifecycle events reported by your SDK to the appropriate `PartnerAdDelegate` methods.
```

> 12. Refer to the Helium SDK’s PartnerAdapter+Log.swift and PartnerAd+Log.swift for a complete list of log events to be used, and to PartnerAdapter+Error.swift and PartnerAd+Error.swift for a complete list of errors

Reviewer: ```suggestion
12. Refer to the Chartboost Mediation SDK’s PartnerAdapter+Log.swift and PartnerAd+Log.swift for a complete list of log events to be used, and to PartnerAdapter+Error.swift and PartnerAd+Error.swift for a complete list of errors
```

TODO LINK

> 13. On the Helium web dashboard, add your full adapter class name so your mediation adapter and partner SDK can be initialized and interacted with for ad serving purposes.

Reviewer: ```suggestion
13. On the Chartboost Mediation web dashboard, add your full adapter class name so your mediation adapter and partner SDK can be initialized and interacted with for ad serving purposes.
```

pod 'ChartboostMediationAdapterReference'
```

## Chartboost Mediation Custom Adapter Implementation Guidelines

> [!IMPORTANT]
> Chartboost Mediation will not be providing official support for any custom adapters. For official adapters that we develop and support, [visit this site](https://adapters.chartboost.com).

An adapter must expose a class that conforms to the `PartnerAdapter` protocol. `PartnerAdapter` is [defined in the Mediation SDK documentation](https://reference.chartboost.com/mediation/ios/4.6.0/documentation/chartboostmediationsdk/), and thoroughly explained in code comments throughout the reference adapter. This README provides additional context.
<br>
<br>

### PartnerAdapter protocol

> #### adatperVersion

Reviewer 2: ```suggestion
#### adapterVersion
```

pod 'ChartboostMediationAdapterReference'
```

## Chartboost Mediation Custom Adapter Implementation Guide

> [!IMPORTANT]
> Chartboost Mediation will not be providing official support for any custom adapters. For official adapters that we develop and support, [visit this site](https://adapters.chartboost.com).

> 1. Create a new class that conforms to Helium’s `PartnerAdapter` protocol.

Reviewer: ```suggestion
1. Create a new class that conforms to Chartboost Mediation's `PartnerAdapter` protocol.
```

> 2. Implement `var partnerSDKVersion: String { get }` with the version of the partner SDK. Most adapters fetch this from the partner SDK's API so the adapter always reports the correct version even if the SKD version has changed.

Reviewer: ```suggestion
2. Implement `var partnerSDKVersion: String { get }` with the version of the partner SDK. Most adapters fetch this from the partner SDK's API so the adapter always reports the correct version even if the SDK version has changed.
```

If a call to this method times out, the Mediation SDK will consider your adapter un-initialized even if you later report a successful init.
9. Implement `func fetchBidderInformation( request: PreBidRequest, completion: @escaping ([String: String]?) -> Void )`. If you support bidding, return `completion([String: String])` of biddable token Strings. The keys used in this dictionary depend on how your network's bidder was integrated with our backend. Usually Chartboost just passes these labels through to the bidding server, so whatever your RTB spec says probably applies here i.e. if your server expects bidder information in a field called 'token', then 'token' is probably the correct dictionary key to use here.  
As you fetch/genereate the token(s), use these three `PartnerLogEvent`s to log execution:  
`.fetchBidderInfoStarted(PreBidRequest)`  
`.fetchBidderInfoSucceeded(PreBidRequest)`  
`.fetchBidderInfoFailed(PreBidRequest, error: Error)`  

> If bidding is not supported by your SDK, just return an empty Dictionary.

Reviewer: ```suggestion
If bidding is not supported by your SDK, just call `completion(nil)`.
```

    
    Each `PartnerAd` also needs to store the `PartnerAdDelegate` that Chartboost Mediation SDK will pass to `makeAd`, and call the following delegate methods to report your ad lifecycle events:  
    a. `func didTrackImpression(_ ad: PartnerAd, details: PartnerEventDetails)`  
    Call when the partner SDK registers an impression for the currently showing ad.  
    b. `func didClick(_ ad: PartnerAd, details: PartnerEventDetails)`  
    Call when the partner ad has been clicked as the result of a user action.  
    c. `func didReward(_ ad: PartnerAd, details: PartnerEventDetails)`  
    Call when a reward has been given for watching a video ad.  
    d. `func didDismiss(_ ad: PartnerAd, details: PartnerEventDetails, error: Error?)`  
    Call when the partner ad has been dismissed as the result of a user action.  
    e. `func didExpire(_ ad: PartnerAd, details: PartnerEventDetails)`  
    Call when the partner ad has expired as determined by the partner SDK.  
11. Implement `func makeAd(request: PartnerAdLoadRequest, delegate: PartnerAdDelegate) throws -> PartnerAd`, which returns your `PartnerAd`-conforming objects.

> If your adapter supports Adaptive Banners or Rewarded Interstitials, you'll need to check for those within the `default` clause, to maintain backward compatibility.

Reviewer: ```suggestion
If you are writing an adapter for Chartboost Mediation 4.x, and your adapter supports Adaptive Banners or Rewarded Interstitials, you'll need to check for those within the `default` clause, to maintain backward compatibility.
```

```swift
default:
    if request.format.rawValue == "rewarded_interstitial" {
        return YourAdapterRewardedInterstitialAd(adapter: self, request: request, delegate: delegate)
    } else if request.format.rawValue == "adaptive_banner" {
        return YourAdapterAdapterBannerAd(adapter: self, request: request, delegate: delegate)
    } else {
        throw error(.loadFailureUnsupportedAdFormat)
    }
```

> 12. On the Helium web dashboard, add your full adapter class name so your mediation adapter and partner SDK can be initialized and interacted with for ad serving purposes.  

Reviewer: ```suggestion
12. On the Chartboost Mediation web dashboard, add your full adapter class name so your mediation adapter and partner SDK can be initialized and interacted with for ad serving purposes.  
```
